### PR TITLE
accouting subquery's cost into total cost  #72 

### PR DESCRIPTION
**Pull request by 9DemonFox**

### CODE REVIEW

File: qpmodel/ClassDiagram1.cd [hunk 1]
```diff
@@ -0,0 +1,1825 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (zhouqingqing) [inline qpmodel/ClassDiagram1.cd:1]: What is this file?

File: qpmodel/codegen.cs [hunk 2]
```diff
@@ -42,7 +42,7 @@
 // Code Generation Targets:
 //  1. specialization with all possible without breaking code structure.
 //  2. can be integrated gradually - we can do one iterator by another.
-//    2.1 expr is not included
+//    2.1 expr is not included+
```
Comment (zhouqingqing) [inline qpmodel/codegen.cs:45]: please remove this accidental change

File: test/UnitTest.cs [hunk 3]
```diff
@@ -384,6 +384,67 @@ public void TestTpchDistributed()
             }
         }
 
+        [TestMethod]
```
Comment (zhouqingqing) [inline test/UnitTest.cs:387]: why do we need an extra test for this fix?

File: qpmodel/PhysicalNode.cs [hunk 4]
```diff
@@ -173,7 +173,21 @@ public double InclusiveCost()
                 else
                     incCost += x.InclusiveCost();
             });
-
+            //FIXED  @issue72 here should add the cost of subquery
+            if (this.logic_.filter_ != null)
```
Comment (zhouqingqing) [inline qpmodel/PhysicalNode.cs:177]: Subquery can show up in selection, filter and some other places - so only handle filter is not enough.

File: qpmodel/ClassDiagram1.cd (hunk 1 above)
Comment (9DemonFox) [reply qpmodel/ClassDiagram1.cd:1]: This is a UML generated by VS , I use it to understand the code,  I will use. gitignore to delete it.

File: qpmodel/PhysicalNode.cs [hunk 5]
```diff
@@ -173,7 +173,21 @@ public double InclusiveCost()
...
+            {
+                Expr expr = this.logic_.filter_;
+                expr.VisitEachT<SubqueryExpr>(x =>
+                {
+                    if (x.query_.physicPlan_ != null)
```
Comment (zhouqingqing) [inline qpmodel/PhysicalNode.cs:182]: Do you have a test case x.query_.physicPlan_ is null? That's a bug shall be fixed (you can open an issue and fix separately).

File: qpmodel/PhysicalNode.cs (hunk 4 above)
Comment (zhouqingqing) [reply qpmodel/PhysicalNode.cs:177]: let's first get filter case fixed and checkin other cases (selection etc) in a separate patch as it is more extensive change.

File: qpmodel/PhysicalNode.cs (hunk 5 above)
Comment (9DemonFox) [reply qpmodel/PhysicalNode.cs:182]: OK. And the new issue is #230

File: qpmodel/codegen.cs (hunk 2 above)
Comment (9DemonFox) [reply qpmodel/codegen.cs:45]: Sorry, I will check carefully next time.

File: test/UnitTest.cs (hunk 3 above)
Comment (9DemonFox) [reply test/UnitTest.cs:387]: I delete the unittest. q20 of TCP-H is more complex for debugging. So I construct a easy one for this issue.
I have renamed the TestForIssue72 to TestSubqueryInclusiveCost and move it to class Subquery in UnitTest.cs.

File: qpmodel/PhysicalNode.cs (hunk 4 above)
Comment (9DemonFox) [reply qpmodel/PhysicalNode.cs:177]: OK. 
And I have find this SQL with uncorrealted suquery
"select c_name,(select count(O.o_orderkey) from orders O where O.o_custkey=37) as OrderCount from customer C"
has a wrong inccost.
image

File: test/UnitTest.cs (hunk 3 above)
Comment (zhouqingqing) [reply test/UnitTest.cs]: I am not sure I got it - the original tpch test already have q20. Why you introduce TestSubqueryInclusiveCost() just to test q20 again?

File: qpmodel/PhysicalNode.cs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-﻿/*
+﻿ /*
```
Comment (zhouqingqing) [inline qpmodel/PhysicalNode.cs:1]: extra ' '

File: test/UnitTest.cs (hunk 3 above)
Comment (9DemonFox) [reply test/UnitTest.cs]: The query result you quoted  in the issue description is the  result  of  q20. I test q20 for emphasis the issue of bug in filter subquery.  If you think it is unnecessary to creat a unit test  to test q20 again, i will delete it.

Comment (9DemonFox) [reply test/UnitTest.cs]: The query result you quoted  in the issue description is the  result  of  q20. I test q20 for emphasis the issue of bug in filter subquery.  If you think it is unnecessary to creat a unit test  to test q20 again, i will delete it.

Comment (9DemonFox) [reply test/UnitTest.cs]: The query result you quoted  in the issue description is the  result  of  q20. I test q20 for emphasis the issue of bug in filter subquery.  If you think it is unnecessary to creat a unit test  to test q20 again, i will delete it.

File: qpmodel/PhysicalNode.cs (hunk 6 above)
Comment (9DemonFox) [reply qpmodel/PhysicalNode.cs:1]: sorry，I will soon  amend it

File: test/UnitTest.cs (hunk 3 above)
Comment (zhouqingqing) [reply test/UnitTest.cs]: please remove this unittest as it is redundant.

**Comment by zhouqingqing** [comment]
Thanks, merged!

**Comment by 9DemonFox** [comment]
Thanks for your advises. And I am not very familiar with details of git, so I have made some very low-level mistake.  Thank you again.

**Pull request merged**